TestEndpoints is flaky

### What did you do?

TestEndpoints shouldn't fail.

### What did you expect to see?

_No response_

### What did you see instead? Under which circumstances?

Got failures twice:

```
--- FAIL: TestEndpoints (0.06s)
    --- FAIL: TestEndpoints/local (0.02s)
        --- FAIL: TestEndpoints/local/run_27_series_"limit=1&match%5B%5D=test_metric1" (0.00s)
            --- FAIL: TestEndpoints/local/run_27_series_"limit=1&match%5B%5D=test_metric1"/GET (0.00s)
                api_test.go:2897: 
                    	Error Trace:	/__w/prometheus/prometheus/util/testutil/cmp.go:40
                    	            				/__w/prometheus/prometheus/util/testutil/cmp.go:29
                    	            				/__w/prometheus/prometheus/web/api/v1/api_test.go:2930
                    	            				/__w/prometheus/prometheus/web/api/v1/api_test.go:2897
                    	Error:      	Not equal: 
                    	            	expected: [{__name__="test_metric1", foo="bar"}]
                    	            	actual  : [{__name__="test_metric1", foo="boo"}]  []labels.Labels{
                    	            	- 	s`{__name__="test_metric1", foo="bar"}`,
                    	            	+ 	s`{__name__="test_metric1", foo="boo"}`,
                    	            	  }
                    	Test:       	TestEndpoints/local/run_27_series_"limit=1&match%5B%5D=test_metric1"/GET
            --- FAIL: TestEndpoints/local/run_27_series_"limit=1&match%5B%5D=test_metric1"/POST (0.00s)
                api_test.go:2897: 
                    	Error Trace:	/__w/prometheus/prometheus/util/testutil/cmp.go:40
                    	            				/__w/prometheus/prometheus/util/testutil/cmp.go:29
                    	            				/__w/prometheus/prometheus/web/api/v1/api_test.go:2930
                    	            				/__w/prometheus/prometheus/web/api/v1/api_test.go:2897
                    	Error:      	Not equal: 
                    	            	expected: [{__name__="test_metric1", foo="bar"}]
                    	            	actual  : [{__name__="test_metric1", foo="boo"}]  []labels.Labels{
                    	            	- 	s`{__name__="test_metric1", foo="bar"}`,
                    	            	+ 	s`{__name__="test_metric1", foo="boo"}`,
                    	            	  }
                    	Test:       	TestEndpoints/local/run_27_series_"limit=1&match%5B%5D=test_metric1"/POST
    --- FAIL: TestEndpoints/remote (0.03s)
        --- FAIL: TestEndpoints/remote/run_27_series_"limit=1&match%5B%5D=test_metric1" (0.00s)
            --- FAIL: TestEndpoints/remote/run_27_series_"limit=1&match%5B%5D=test_metric1"/GET (0.00s)
                api_test.go:2897: 
                    	Error Trace:	/__w/prometheus/prometheus/util/testutil/cmp.go:40
                    	            				/__w/prometheus/prometheus/util/testutil/cmp.go:29
                    	            				/__w/prometheus/prometheus/web/api/v1/api_test.go:2930
                    	            				/__w/prometheus/prometheus/web/api/v1/api_test.go:2897
                    	Error:      	Not equal: 
                    	            	expected: [{__name__="test_metric1", foo="bar"}]
                    	            	actual  : [{__name__="test_metric1", foo="boo"}]  []labels.Labels{
                    	            	- 	s`{__name__="test_metric1", foo="bar"}`,
                    	            	+ 	s`{__name__="test_metric1", foo="boo"}`,
                    	            	  }
                    	Test:       	TestEndpoints/remote/run_27_series_"limit=1&match%5B%5D=test_metric1"/GET
            --- FAIL: TestEndpoints/remote/run_27_series_"limit=1&match%5B%5D=test_metric1"/POST (0.00s)
                api_test.go:2897: 
                    	Error Trace:	/__w/prometheus/prometheus/util/testutil/cmp.go:40
                    	            				/__w/prometheus/prometheus/util/testutil/cmp.go:29
                    	            				/__w/prometheus/prometheus/web/api/v1/api_test.go:2930
                    	            				/__w/prometheus/prometheus/web/api/v1/api_test.go:2897
                    	Error:      	Not equal: 
                    	            	expected: [{__name__="test_metric1", foo="bar"}]
                    	            	actual  : [{__name__="test_metric1", foo="boo"}]  []labels.Labels{
                    	            	- 	s`{__name__="test_metric1", foo="bar"}`,
                    	            	+ 	s`{__name__="test_metric1", foo="boo"}`,
                    	            	  }
                    	Test:       	TestEndpoints/remote/run_27_series_"limit=1&match%5B%5D=test_metric1"/POST
FAIL
FAIL	github.com/prometheus/prometheus/web/api/v1	0.136s
FAIL
```

### System information

CI

### Prometheus version

```text
0794ed76ddc47e1577e430b99ed81aa17c067607
```


### Prometheus configuration file

_No response_

### Alertmanager version

_No response_

### Alertmanager configuration file

_No response_

### Logs

```text
https://github.com/prometheus/prometheus/actions/runs/8145220860/job/22260980928?pr=12532
```

--- a/web/api/v1/api_test.go
+++ b/web/api/v1/api_test.go
@@ -1054,7 +1054,7 @@ func testEndpoints(t *testing.T, api *API, tr *testTargetRetriever, es storage.E
 		params                map[string]string
 		query                 url.Values
 		response              interface{}
-		responseLen           int
+		responseLen           int // If nonzero, check only the length; `response` is ignored.
 		responseMetadataTotal int
 		responseAsJSON        string
 		errType               errorType
@@ -1388,17 +1388,16 @@ func testEndpoints(t *testing.T, api *API, tr *testTargetRetriever, es storage.E
 				labels.FromStrings("__name__", "test_metric2", "foo", "boo"),
 			},
 		},
-		// Missing match[] query params in series requests.
+		// Series request with limit.
 		{
 			endpoint: api.series,
 			query: url.Values{
 				"match[]": []string{"test_metric1"},
 				"limit":   []string{"1"},
 			},
-			response: []labels.Labels{
-				labels.FromStrings("__name__", "test_metric1", "foo", "bar"),
-			},
+			responseLen: 1, // API does not specify which particular value will come back.
 		},
+		// Missing match[] query params in series requests.
 		{
 			endpoint: api.series,
 			errType:  errorBadData,
@@ -2670,18 +2669,16 @@ func testEndpoints(t *testing.T, api *API, tr *testTargetRetriever, es storage.E
 					"boo",
 				},
 			},
+			// Label values with limit.
 			{
 				endpoint: api.labelValues,
 				params: map[string]string{
-					"name": "foo",
+					"name": "__name__",
 				},
 				query: url.Values{
-					"match[]": []string{"test_metric4"},
-					"limit":   []string{"1"},
+					"limit": []string{"2"},
 				},
-				response: []string{
-					"bar",
-				},
+				responseLen: 2, // API does not specify which particular values will come back.
 			},
 			// Label names.
 			{
@@ -2822,13 +2819,13 @@ func testEndpoints(t *testing.T, api *API, tr *testTargetRetriever, es storage.E
 				},
 				response: []string{"__name__", "foo"},
 			},
+			// Label names with limit.
 			{
 				endpoint: api.labelNames,
 				query: url.Values{
-					"match[]": []string{"test_metric2"},
-					"limit":   []string{"1"},
+					"limit": []string{"2"},
 				},
-				response: []string{"__name__"},
+				responseLen: 2, // API does not specify which particular values will come back.
 			},
 		}...)
 	}

